File: game/curses.py
# -*- coding: utf-8 -*-
"""Oferece, no momento certo, a decisão de ativar uma Maldição virada para
baixo em reação a um evento específico — espelha webgame/curses.js.

GAME_DESIGN.md: "Revelar/ativar uma Maldição já setada: a qualquer momento
no turno do OPONENTE." O jogo nunca revela uma Maldição sozinho: sempre que
um evento bate com o gatilho declarado (EFFECTS.registrar_gatilho_maldicao,
game/effects.py) de alguma Maldição virada pra baixo de quem NÃO está na
vez, um pedido de seleção pergunta pro dono se quer ativá-la agora — o motor
espera essa decisão (mesmo mecanismo de `ctrl.solicitar_selecao` de sempre)
antes de continuar."""
from __future__ import annotations

from typing import Callable

from .components import CardInfo, FaceDown, ManaCost


def maldicoes_elegiveis(ctrl, evento) -> list[tuple[int, int]]:
    """Devolve [(dono_id, card), ...] entre as Maldições viradas pra baixo de
    quem NÃO está na vez, com um gatilho declarado batendo com o TIPO do
    evento, cuja condição aceita, e que o dono ainda tem mana pra pagar."""
    elegiveis: list[tuple[int, int]] = []
    ts = ctrl.fase_atual()
    for pid in ctrl.jogadores:
        if pid == ts.jogador_da_vez:
            continue  # só quem NÃO está na vez pode reagir
        lado = ctrl.board.lado(pid)
        for card in lado.magia:
            if card is None or not ctrl.world.has_component(card, FaceDown):
                continue
            custo = ctrl.world.get_component(card, ManaCost).valor
            if custo > ctrl.players[pid].mana:
                continue
            info = ctrl.world.get_component(card, CardInfo)
            gatilho = ctrl.effects.gatilho_maldicao(info.nome)
            if gatilho is None:
                continue
            evento_tipo, condicao = gatilho
            if evento_tipo is not type(evento):
                continue
            if condicao(evento, ctrl, pid):
                elegiveis.append((pid, card))
    return elegiveis


def ofertar_maldicoes_reativas(ctrl, evento, continuar: Callable[[], None] = lambda: None) -> None:
    """Pausa o fluxo — mesmo `ctrl.solicitar_selecao` de sempre — até que
    cada dono elegível decida ativar uma Maldição (ou nenhuma). `continuar`
    só roda depois de TODAS as decisões resolvidas (na prática, quase
    sempre um único dono)."""
    from .actions import AcaoInvalida, ActivateSetCurseAction  # import tardio: evita ciclo com actions.py

    elegiveis = maldicoes_elegiveis(ctrl, evento)
    donos = list(dict.fromkeys(pid for pid, _ in elegiveis))

    def processar(i: int) -> None:
        if i >= len(donos):
            continuar()
            return
        dono_id = donos[i]
        lado = ctrl.board.lado(dono_id)
        cartas = [c for pid, c in elegiveis
                  if pid == dono_id and c in lado.magia and ctrl.world.has_component(c, FaceDown)]
        if not cartas:
            processar(i + 1)
            return

        def _ao_escolher(escolha: list[int]) -> None:
            if escolha:
                try:
                    ActivateSetCurseAction(player_id=dono_id, card=escolha[0], evento=evento).executar(ctrl)
                except AcaoInvalida:
                    pass  # defensivo: estado pode ter mudado entre a oferta e a escolha
            processar(i + 1)

        ctrl.solicitar_selecao(
            player_id=dono_id,
            prompt="Uma Maldição virada para baixo pode ser ativada agora. Ativar uma delas?",
            opcoes=cartas,
            on_resolved=_ao_escolher,
            minimo=0, maximo=1,
        )

    processar(0)

File: webgame/curses.js
// Oferece, no momento certo, a decisão de ativar uma Maldição virada para
// baixo em reação a um evento específico — GAME_DESIGN.md: "Revelar/ativar
// uma Maldição já setada: a qualquer momento no turno do OPONENTE." Cada
// Maldição declara (regGatilhoMaldicao, em effects.js) o evento que a torna
// elegível pro modal; isso substitui a lógica antiga (só existia pro ataque,
// e oferecia QUALQUER Maldição virada pra baixo, sem checar se fazia sentido
// pro evento).
var TCG = window.TCG || (window.TCG = {});

// Acha, entre as Maldições viradas pra baixo de quem NÃO está na vez, as que
// têm um gatilho declarado batendo com `evento.tipo`, cuja condição aceita, e
// que o dono ainda tem mana pra pagar (senão a escolha ia estourar ao tentar
// ativar — mesmo cuidado que a IA já tinha antes de oferecer "revelar").
TCG.maldicoesElegiveis = function maldicoesElegiveis(game, evento) {
  const elegiveis = [];
  for (const pid of game.jogadores) {
    if (pid === game.jogadorDaVez) continue; // só quem NÃO está na vez pode reagir
    for (const carta of game.board[pid].magia) {
      if (!carta || !carta.faceDown) continue;
      if (carta.custoMana > game.players[pid].mana) continue;
      const gatilho = TCG.GATILHOS_DE_MALDICAO[carta.nome];
      if (!gatilho || gatilho.eventoTipo !== evento.tipo) continue;
      if (gatilho.condicao(evento, game, pid)) elegiveis.push({ donoId: pid, carta });
    }
  }
  return elegiveis;
};

// Pausa o fluxo — mesmo mecanismo de seleção que já mostra modal pro jogador
// local e resolve sozinho pra IA (ver selection.js e o listener genérico de
// "selecaoPedida" em ui.js) — até que cada dono elegível decida ativar uma
// Maldição (ou nenhuma). `continuar` só roda depois de TODAS as decisões
// resolvidas (na prática, quase sempre um único dono).
TCG.ofertarMaldicoesReativas = function ofertarMaldicoesReativas(game, evento, continuar = () => {}) {
  const elegiveis = TCG.maldicoesElegiveis(game, evento);
  const donos = [...new Set(elegiveis.map((e) => e.donoId))];

  function processar(i) {
    if (i >= donos.length) { continuar(); return; }
    const donoId = donos[i];
    const lado = game.board[donoId];
    const cartas = elegiveis
      .filter((e) => e.donoId === donoId)
      .map((e) => e.carta)
      .filter((c) => lado.magia.includes(c) && c.faceDown); // ainda válidas nesse instante
    if (!cartas.length) { processar(i + 1); return; }
    game.selection.solicitar(
      donoId,
      "Uma Maldição virada para baixo pode ser ativada agora. Ativar uma delas?",
      cartas,
      (escolha) => {
        if (escolha.length) {
          try {
            TCG.acoes.ativarMaldicaoSetada(game, donoId, escolha[0], evento);
          } catch (e) {
            if (!(e instanceof TCG.AcaoInvalida)) throw e; // defensivo: estado pode ter mudado entre a oferta e a escolha
          }
        }
        processar(i + 1);
      },
      0, 1
    );
  }
  processar(0);
};
